File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_CSharp/SolverGenerator/Techniques/HiddenSubsetTechnique.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: HiddenSubsetTechnique.cs
//
//  Description: Implements the Sudoku hidden subset technique.
// 
//--------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections;

namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
{
	/// <summary>Implements the hidden pair elimination technique.</summary>
	[Serializable]
    public sealed class HiddenPairTechnique : HiddenSubsetTechnique
	{
		/// <summary>Initializes the technique.</summary>
		public HiddenPairTechnique() : base(2) {}

		/// <summary>Gets the difficulty level of this technique.</summary>
		internal override uint DifficultyLevel { get { return 6; } }
	}

	/// <summary>Implements the hidden triplet elimination technique.</summary>
	[Serializable]
    public sealed class HiddenTripletTechnique : HiddenSubsetTechnique
	{
		/// <summary>Initializes the technique.</summary>
		public HiddenTripletTechnique() : base(3) {}

		/// <summary>Gets the difficulty level of this technique.</summary>
		internal override uint DifficultyLevel { get { return 8; } }
	}

	/// <summary>Implements the hidden quad elimination technique.</summary>
	[Serializable]
    public sealed class HiddenQuadTechnique : HiddenSubsetTechnique
	{
		/// <summary>Initializes the technique.</summary>
		public HiddenQuadTechnique() : base(4) {}

		/// <summary>Gets the difficulty level of this technique.</summary>
		internal override uint DifficultyLevel { get { return 10; } }
	}

	/// <summary>Implements the hidden subset elimination technique.</summary>
	[Serializable]
    public abstract class HiddenSubsetTechnique : EliminationTechnique
	{
		/// <summary>The size of the subset to evaluate.</summary>
		private int _subsetSize;
		private int [] _foundLocations;

		/// <summary>Initialize the technique.</summary>
		/// <param name="subsetSize">The size of the subset to evaluate.</param>
		protected HiddenSubsetTechnique(int subsetSize)
		{
			if (subsetSize < 2) throw new ArgumentOutOfRangeException("subsetSize");
			_subsetSize = subsetSize;
			_foundLocations = new int[_subsetSize]; // optimization, rather than allocating on each call
		}

        public override bool Equals(object obj)
        {
            HiddenSubsetTechnique hst = obj as HiddenSubsetTechnique;
            if (hst == null) return false;
            return _subsetSize == hst._subsetSize;
        }

        public override int GetHashCode()
        {
            return GetType().GetHashCode();
        }

        public override EliminationTechnique Clone()
        {
            HiddenSubsetTechnique hst = (HiddenSubsetTechnique)base.Clone();
            hst._foundLocations = new int[_subsetSize];
            return hst;
        }

		/// <summary>Performs hidden subset elimination on one dimension of possible numbers.</summary>
		/// <param name="possibleNumbers">The row/column/box to analyze.</param>
		/// <returns>The number of changes that were made to the possible numbers.</returns>
		private int EliminateHiddenSubsets(FastBitArray [] possibleNumbers,
			bool exitEarlyWhenSoleFound, out bool exitedEarly)
		{
			int changesMade = 0;
			exitedEarly = false;
			int numLocations;
			int [] foundLocations = _foundLocations; // optimization, rather than allocating on each call

			// We'll look starting with each cell in the row/column/box
			for(int i=0; i<possibleNumbers.Length; i++)
			{
				// Only look at the cell if it has at least subsetSize values set,
				// otherwise it can't be part of a hidden subset
				int numPossible = possibleNumbers[i].CountSet;
				if (numPossible >= _subsetSize)
				{
					// For each combination
					foreach(int [] combination in CreateCombinations(_subsetSize, possibleNumbers[i].GetSetBits()))
					{
						// Find other cells that contain that same combination,
						// but only up to the subset size
						numLocations = 0;
						foundLocations[numLocations++] = i;
						for(int j=i+1; j<possibleNumbers.Length && numLocations < foundLocations.Length; j++)
						{
							if (AllAreSet(combination, possibleNumbers[j]))
							{
								foundLocations[numLocations++] = j;
							}
						}

						if (numLocations == foundLocations.Length)
						{
							bool isValidHidden = true;

							// Make sure that none of the numbers appear in any other cell
							for(int j=0; j<possibleNumbers.Length && isValidHidden; j++)
							{
								bool isFoundLocation = Array.BinarySearch(foundLocations, j) >= 0;
								if (!isFoundLocation && AnyAreSet(combination, possibleNumbers[j]))
								{
									isValidHidden = false;
									break;
								}
							}

							// If this is a valid hidden subset, eliminate all other numbers
							// from each cell in the subset
							if (isValidHidden)
							{
								foreach(int foundLoc in foundLocations)
								{
									FastBitArray possibleNumbersForLoc = possibleNumbers[foundLoc];
									foreach(int n in possibleNumbersForLoc.GetSetBits())
									{
										if (Array.BinarySearch(combination, n) < 0)
										{
											possibleNumbersForLoc[n] = false;
											changesMade++;
										}
									}
									if (exitEarlyWhenSoleFound &&
										possibleNumbersForLoc.CountSet == 1)
									{
										exitedEarly = true;
										return changesMade;
									}
								}
								break;
							}
						}
					}
				}
			}

			return changesMade;
		}

        /// <summary>Allows for iteration through all combinations from an array.</summary>
        /// <param name="size">The size of the subset.</param>
        /// <param name="numbers">The numbers.</param>
        /// <returns>An enumerable for all the combinations.</returns>
        public static IEnumerable<int[]> CreateCombinations(int size, int[] numbers)
        {
            int[] indices = new int[size];
            int[] result = new int[size];
            for (int i = 0; i < indices.Length; i++)
            {
                indices[i] = i;
                result[i] = numbers[i];
            }
            yield return result;

            while (true)
            {
                if (indices[0] >= numbers.Length - size) yield break;

                // Find the next position to be incremented
                int pos;
                for (pos = indices.Length - 1;
                    pos > 0 && indices[pos] == numbers.Length - size + pos;
                    pos--) ;

                // Increment it
                indices[pos]++;

                // Increment everything else
                for (int j = pos; j < size - 1; j++) indices[j + 1] = indices[j] + 1;

                // Create the result
                for (int i = 0; i < indices.Length; i++) result[i] = numbers[indices[i]];
                yield return result;
            }
        }

		/// <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		/// <param name="state">The puzzle state.</param>
		/// <param name="exitEarlyWhenSoleFound">Whether the method can exit early when a cell with only one possible number is found.</param>
		/// <param name="possibleNumbers">The previously computed possible numbers.</param>
		/// <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		/// <param name="exitedEarly">Whether the method exited early due to a cell with only one value being found.</param>
		/// <returns>Whether more changes may be possible based on changes made during this execution.</returns>
		internal override bool Execute(
			PuzzleState state, bool exitEarlyWhenSoleFound,
			FastBitArray[][] possibleNumbers, out int numberOfChanges, out bool exitedEarly)
		{
			numberOfChanges = 0;
			exitedEarly = false;
			FastBitArray [] arrays = new FastBitArray[state.GridSize];

			for(int i=0; i<state.GridSize; i++)
			{
				GetRowPossibleNumbers(possibleNumbers, i, arrays);
				numberOfChanges += EliminateHiddenSubsets(arrays, exitEarlyWhenSoleFound, out exitedEarly);
				if (exitedEarly) return false;

				GetColumnPossibleNumbers(possibleNumbers, i, arrays);
				numberOfChanges += EliminateHiddenSubsets(arrays, exitEarlyWhenSoleFound, out exitedEarly);
				if (exitedEarly) return false;

				GetBoxPossibleNumbers(state, possibleNumbers, i, arrays);
				numberOfChanges += EliminateHiddenSubsets(arrays, exitEarlyWhenSoleFound, out exitedEarly);
				if (exitedEarly) return false;
			}

			return numberOfChanges != 0;
		}
	}
}

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_VisualBasic/SolverGenerator/Techniques/HiddenSubsetTechnique.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: HiddenSubsetTechnique.vb
'
'  Description: Implements the Sudoku hidden subset technique.
' 
'--------------------------------------------------------------------------

Imports Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections

Namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
	''' <summary>Implements the hidden pair elimination technique.</summary>
	<Serializable>
	Public NotInheritable Class HiddenPairTechnique
		Inherits HiddenSubsetTechnique
		''' <summary>Initializes the technique.</summary>
		Public Sub New()
			MyBase.New(2)
		End Sub

		''' <summary>Gets the difficulty level of this technique.</summary>
		Friend Overrides ReadOnly Property DifficultyLevel() As UInteger
			Get
				Return 6
			End Get
		End Property
	End Class

	''' <summary>Implements the hidden triplet elimination technique.</summary>
	<Serializable>
	Public NotInheritable Class HiddenTripletTechnique
		Inherits HiddenSubsetTechnique
		''' <summary>Initializes the technique.</summary>
		Public Sub New()
			MyBase.New(3)
		End Sub

		''' <summary>Gets the difficulty level of this technique.</summary>
		Friend Overrides ReadOnly Property DifficultyLevel() As UInteger
			Get
				Return 8
			End Get
		End Property
	End Class

	''' <summary>Implements the hidden quad elimination technique.</summary>
	<Serializable>
	Public NotInheritable Class HiddenQuadTechnique
		Inherits HiddenSubsetTechnique
		''' <summary>Initializes the technique.</summary>
		Public Sub New()
			MyBase.New(4)
		End Sub

		''' <summary>Gets the difficulty level of this technique.</summary>
		Friend Overrides ReadOnly Property DifficultyLevel() As UInteger
			Get
				Return 10
			End Get
		End Property
	End Class

	''' <summary>Implements the hidden subset elimination technique.</summary>
	<Serializable>
	Public MustInherit Class HiddenSubsetTechnique
		Inherits EliminationTechnique
		''' <summary>The size of the subset to evaluate.</summary>
		Private _subsetSize As Integer
		Private _foundLocations() As Integer

		''' <summary>Initialize the technique.</summary>
		''' <param name="subsetSize">The size of the subset to evaluate.</param>
		Protected Sub New(ByVal subsetSize As Integer)
			If subsetSize < 2 Then
				Throw New ArgumentOutOfRangeException("subsetSize")
			End If
			_subsetSize = subsetSize
			_foundLocations = New Integer(_subsetSize - 1){} ' optimization, rather than allocating on each call
		End Sub

		Public Overrides Overloads Function Equals(ByVal obj As Object) As Boolean
            Dim hst = TryCast(obj, HiddenSubsetTechnique)
			If hst Is Nothing Then
				Return False
			End If
			Return _subsetSize = hst._subsetSize
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return Me.GetType().GetHashCode()
		End Function

		Public Overrides Function Clone() As EliminationTechnique
            Dim hst = CType(MyBase.Clone(), HiddenSubsetTechnique)
			hst._foundLocations = New Integer(_subsetSize - 1){}
			Return hst
		End Function

		''' <summary>Performs hidden subset elimination on one dimension of possible numbers.</summary>
		''' <param name="possibleNumbers">The row/column/box to analyze.</param>
		''' <returns>The number of changes that were made to the possible numbers.</returns>
        Private Function EliminateHiddenSubsets(ByVal possibleNumbers() As FastBitArray,
             ByVal exitEarlyWhenSoleFound As Boolean, <System.Runtime.InteropServices.Out()> ByRef exitedEarly As Boolean) As Integer
            Dim changesMade = 0
            exitedEarly = False
            Dim numLocations As Integer
            Dim foundLocations() = _foundLocations ' optimization, rather than allocating on each call

            ' We'll look starting with each cell in the row/column/box
            For i = 0 To possibleNumbers.Length - 1
                ' Only look at the cell if it has at least subsetSize values set,
                ' otherwise it can't be part of a hidden subset
                Dim numPossible = possibleNumbers(i).CountSet
                If numPossible >= _subsetSize Then
                    ' For each combination
                    For Each combination In CreateCombinations(_subsetSize, possibleNumbers(i).GetSetBits())
                        ' Find other cells that contain that same combination,
                        ' but only up to the subset size
                        numLocations = 0
                        foundLocations(numLocations) = i
                        numLocations += 1
                        Dim j = i + 1
                        Do While j < possibleNumbers.Length AndAlso numLocations < foundLocations.Length
                            If AllAreSet(combination, possibleNumbers(j)) Then
                                foundLocations(numLocations) = j
                                numLocations += 1
                            End If
                            j += 1
                        Loop


                        If numLocations = foundLocations.Length Then
                            Dim isValidHidden = True

                            ' Make sure that none of the numbers appear in any other cell
                            j = 0
                            Do While j < possibleNumbers.Length AndAlso isValidHidden
                                Dim isFoundLocation = Array.BinarySearch(foundLocations, j) >= 0
                                If (Not isFoundLocation) AndAlso AnyAreSet(combination, possibleNumbers(j)) Then
                                    isValidHidden = False
                                    Exit Do
                                End If
                                j += 1
                            Loop

                            ' If this is a valid hidden subset, eliminate all other numbers
                            ' from each cell in the subset
                            If isValidHidden Then
                                For Each foundLoc In foundLocations
                                    Dim possibleNumbersForLoc = possibleNumbers(foundLoc)
                                    For Each n In possibleNumbersForLoc.GetSetBits()
                                        If Array.BinarySearch(combination, n) < 0 Then
                                            possibleNumbersForLoc(n) = False
                                            changesMade += 1
                                        End If
                                    Next n
                                    If exitEarlyWhenSoleFound AndAlso possibleNumbersForLoc.CountSet = 1 Then
                                        exitedEarly = True
                                        Return changesMade
                                    End If
                                Next foundLoc
                                Exit For
                            End If
                        End If
                    Next combination
                End If
            Next i

            Return changesMade
        End Function

		''' <summary>Allows for iteration through all combinations from an array.</summary>
		''' <param name="size">The size of the subset.</param>
		''' <param name="numbers">The numbers.</param>
		''' <returns>An enumerable for all the combinations.</returns>
        Public Shared Function CreateCombinations(ByVal size As Integer, ByVal numbers() As Integer) As IEnumerable(Of Integer())
            Dim indices(size - 1) As Integer
            Dim result(size - 1) As Integer
            Dim stringArray(size - 1) As Integer
            Dim pos As Integer = Nothing
            Dim stringvalue As New List(Of Integer())
            For i = 0 To indices.Length - 1
                indices(i) = i
                result(i) = numbers(i)
                stringArray(i) = numbers(i)
            Next i

            Dim strings As New List(Of Integer())
            strings.Add(stringArray)
            Return strings

            Do
                If indices(0) >= numbers.Length - size Then
                    Exit Do
                End If

                ' Find the next position to be incremented
                pos = indices.Length - 1
                Do While pos > 0 AndAlso indices(pos) = numbers.Length - size + pos

                    pos -= 1
                Loop
                ' Increment it
                indices(pos) += 1

                ' Increment everything else
                For j = pos To size - 2
                    indices(j + 1) = indices(j) + 1
                Next j

                ' Create the result
                For i = 0 To indices.Length - 1
                    stringArray(i) = numbers(indices(i))
                Next i
                stringvalue.Add(stringArray)
                Return stringvalue
            Loop
        End Function

		''' <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		''' <param name="state">The puzzle state.</param>
		''' <param name="exitEarlyWhenSoleFound">Whether the method can exit early when a cell with only one possible number is found.</param>
		''' <param name="possibleNumbers">The previously computed possible numbers.</param>
		''' <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		''' <param name="exitedEarly">Whether the method exited early due to a cell with only one value being found.</param>
		''' <returns>Whether more changes may be possible based on changes made during this execution.</returns>
        Friend Overrides Function Execute(ByVal state As PuzzleState, ByVal exitEarlyWhenSoleFound As Boolean,
                    ByVal possibleNumbers()() As FastBitArray, <System.Runtime.InteropServices.Out()>
                    ByRef numberOfChanges As Integer, <System.Runtime.InteropServices.Out()> ByRef exitedEarly As Boolean) As Boolean
            numberOfChanges = 0
            exitedEarly = False
            Dim arrays(state.GridSize - 1) As FastBitArray

            For i = 0 To state.GridSize - 1
                GetRowPossibleNumbers(possibleNumbers, i, arrays)
                numberOfChanges += EliminateHiddenSubsets(arrays, exitEarlyWhenSoleFound, exitedEarly)
                If exitedEarly Then
                    Return False
                End If

                GetColumnPossibleNumbers(possibleNumbers, i, arrays)
                numberOfChanges += EliminateHiddenSubsets(arrays, exitEarlyWhenSoleFound, exitedEarly)
                If exitedEarly Then
                    Return False
                End If

                GetBoxPossibleNumbers(state, possibleNumbers, i, arrays)
                numberOfChanges += EliminateHiddenSubsets(arrays, exitEarlyWhenSoleFound, exitedEarly)
                If exitedEarly Then
                    Return False
                End If
            Next i

            Return numberOfChanges <> 0
        End Function
    End Class
End Namespace